File: solved.ac/CLASS2/2805.py
import sys
input = sys.stdin.readline

def binary():
    left = 0
    right = max(arr)
    val = 0

    while left <= right:
        mid = (left + right) // 2
        tree = 0
        for elem in arr:
            if elem > mid:
                tree += elem - mid

        if tree >= m:
            left = mid + 1
            val = max(val, mid)
        else:
            right = mid - 1
    return val

n, m = map(int, input().split())

arr = list(map(int, input().split()))

print(binary())

File: solved.ac/CLASS2_JS/2805.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const stdin = fs
  .readFileSync(inputPath)
  .toString()
  .split("\n")
  .map((line) => line.trim());

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const solution = () => {
  const [n, m] = input().split(" ").map(Number);
  const arr = input().split(" ").map(Number);

  let [left, right] = [0, Math.max(...arr)];

  while (left <= right) {
    const mid = parseInt((left + right) / 2);
    let temp = 0;

    for (let i = 0; i < arr.length; i++) {
      temp += arr[i] > mid ? arr[i] - mid : 0;
    }

    if (temp >= m) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  console.log(right);
};

solution();
